File: ios/Awaytime/Services/ErrorIntegration.swift
import SwiftUI
import Combine

/// Integration service that connects error handling with other app services
@MainActor
class ErrorIntegration: ObservableObject {
    private let errorService = GlobalErrorHandler.shared.getErrorService()
    private let usageTrackingService = UsageTrackingService()
    private let appBlockingService = AppBlockingService()
    private let goalTrackingService = GoalTrackingService()
    private let coreDataManager = CoreDataManager.shared
    
    private var cancellables = Set<AnyCancellable>()
    
    init() {
        setupErrorIntegration()
        startErrorMonitoring()
    }
    
    private func setupErrorIntegration() {
        // Monitor usage tracking errors
        NotificationCenter.default.publisher(for: .usageTrackingError)
            .sink { [weak self] notification in
                if let error = notification.object as? Error {
                    self?.errorService.handleDeviceActivityError(error)
                }
            }
            .store(in: &cancellables)
        
        // Monitor blocking errors
        NotificationCenter.default.publisher(for: .appBlockingError)
            .sink { [weak self] notification in
                if let userInfo = notification.userInfo,
                   let appName = userInfo["appName"] as? String {
                    self?.errorService.handleBlockingError(appName: appName)
                }
            }
            .store(in: &cancellables)
        
        // Monitor permission changes
        NotificationCenter.default.publisher(for: .permissionStatusChanged)
            .sink { [weak self] notification in
                self?.handlePermissionStatusChange(notification)
            }
            .store(in: &cancellables)
        
        // Monitor data corruption
        NotificationCenter.default.publisher(for: .dataCorruptionDetected)
            .sink { [weak self] _ in
                self?.errorService.handleDataCorruption()
            }
            .store(in: &cancellables)
        
        // Monitor network errors
        NotificationCenter.default.publisher(for: .networkError)
            .sink { [weak self] _ in
                self?.errorService.handleNetworkError()
            }
            .store(in: &cancellables)
    }
    
    private func handlePermissionStatusChange(_ notification: Notification) {
        guard let userInfo = notification.userInfo,
              let permissionType = userInfo["permissionType"] as? String,
              let isGranted = userInfo["isGranted"] as? Bool else { return }
        
        let type: ErrorHandlingService.PermissionType
        switch permissionType {
        case "familyControls":
            type = .familyControls
        case "notifications":
            type = .notifications
        case "screenTime":
            type = .screenTime
        default:
            return
        }
        
        if !isGranted {
            errorService.handlePermissionError(type, denied: false)
        }
    }
    
    private func startErrorMonitoring() {
        errorService.startErrorMonitoring()
        
        // Additional monitoring for specific services
        monitorUsageTracking()
        monitorAppBlocking()
        monitorDataIntegrity()
    }
    
    private func monitorUsageTracking() {
        Timer.scheduledTimer(withTimeInterval: 60, repeats: true) { [weak self] _ in
            self?.checkUsageTrackingHealth()
        }
    }
    
    private func checkUsageTrackingHealth() {
        Task {
            do {
                _ = try await usageTrackingService.getCurrentUsage()
            } catch {
                errorService.handleDeviceActivityError(error)
            }
        }
    }
    
    private func monitorAppBlocking() {
        Timer.scheduledTimer(withTimeInterval: 120, repeats: true) { [weak self] _ in
            self?.checkAppBlockingHealth()
        }
    }
    
    private func checkAppBlockingHealth() {
        Task {
            do {
                try await appBlockingService.validateBlockingStatus()
            } catch {
                errorService.handleFamilyControlsError(error)
            }
        }
    }
    
    private func monitorDataIntegrity() {
        Timer.scheduledTimer(withTimeInterval: 300, repeats: true) { [weak self] _ in
            self?.checkDataIntegrity()
        }
    }
    
    private func checkDataIntegrity() {
        do {
            // Test basic data operations
            _ = try coreDataManager.fetchSelectedApps()
            _ = try coreDataManager.fetchGoals()
            _ = try coreDataManager.fetchUsageRecords(for: Date())
        } catch {
            errorService.handleStorageError(error)
        }
    }
    
    // MARK: - Public Interface
    
    func getErrorService() -> ErrorHandlingService {
        return errorService
    }
    
    func reportError(_ error: Error, context: String) {
        print("🚨 Error reported from \(context): \(error.localizedDescription)")
        errorService.handleUnknownError(error)
    }
    
    func reportPermissionError(_ type: ErrorHandlingService.PermissionType, denied: Bool = true) {
        errorService.handlePermissionError(type, denied: denied)
    }
    
    func reportBlockingFailure(for appName: String) {
        errorService.handleBlockingError(appName: appName)
    }
    
    func reportDataCorruption() {
        errorService.handleDataCorruption()
    }
    
    func reportNetworkError() {
        errorService.handleNetworkError()
    }
}

// MARK: - Error-Aware Service Wrappers

/// Wrapper for UsageTrackingService with error handling
class SafeUsageTrackingService {
    private let usageService = UsageTrackingService()
    private let errorIntegration = ErrorIntegration()
    
    func getCurrentUsage() async -> [AppUsageData] {
        do {
            return try await usageService.getCurrentUsage()
        } catch {
            errorIntegration.reportError(error, context: "UsageTracking")
            return []
        }
    }
    
    func getTodayUsage() -> [AppUsageData] {
        do {
            return usageService.getTodayUsage()
        } catch {
            errorIntegration.reportError(error, context: "UsageTracking")
            return []
        }
    }
    
    func startMonitoring() {
        do {
            try usageService.startMonitoring()
        } catch {
            errorIntegration.reportError(error, context: "UsageTracking")
        }
    }
}

/// Wrapper for AppBlockingService with error handling
class SafeAppBlockingService {
    private let blockingService = AppBlockingService()
    private let errorIntegration = ErrorIntegration()
    
    func blockApps(_ apps: [String]) async {
        do {
            try await blockingService.blockApps(apps)
        } catch {
            // Report specific app blocking failures
            for app in apps {
                errorIntegration.reportBlockingFailure(for: app)
            }
        }
    }
    
    func unblockApps(_ apps: [String]) async {
        do {
            try await blockingService.unblockApps(apps)
        } catch {
            errorIntegration.reportError(error, context: "AppBlocking")
        }
    }
    
    func initializeBlocking() async {
        do {
            try await blockingService.initializeBlocking()
        } catch {
            errorIntegration.reportError(error, context: "AppBlocking")
        }
    }
}

/// Wrapper for GoalTrackingService with error handling
class SafeGoalTrackingService {
    private let goalService = GoalTrackingService()
    private let errorIntegration = ErrorIntegration()
    
    func saveGoal(_ goal: Goal) {
        do {
            try goalService.saveGoal(goal)
        } catch {
            errorIntegration.reportError(error, context: "GoalTracking")
        }
    }
    
    func getCurrentGoals() -> [Goal] {
        do {
            return try goalService.getCurrentGoals()
        } catch {
            errorIntegration.reportError(error, context: "GoalTracking")
            return []
        }
    }
    
    func updateGoalProgress(_ goalId: String, progress: Double) {
        do {
            try goalService.updateGoalProgress(goalId, progress: progress)
        } catch {
            errorIntegration.reportError(error, context: "GoalTracking")
        }
    }
}

// MARK: - Error Recovery Helpers

extension ErrorIntegration {
    
    /// Attempt to recover from common error scenarios
    func attemptAutoRecovery(for error: ErrorHandlingService.AwayTimeError) -> Bool {
        switch error {
        case .blockingFailed:
            return attemptBlockingRecovery()
        case .deviceActivityError, .familyControlsError:
            return attemptSystemServiceRecovery()
        case .storageError:
            return attemptStorageRecovery()
        default:
            return false
        }
    }
    
    private func attemptBlockingRecovery() -> Bool {
        Task {
            do {
                try await appBlockingService.initializeBlocking()
                errorService.clearError()
                return true
            } catch {
                return false
            }
        }
        return false
    }
    
    private func attemptSystemServiceRecovery() -> Bool {
        Task {
            do {
                // Try to restart monitoring
                try usageTrackingService.startMonitoring()
                errorService.clearError()
                return true
            } catch {
                return false
            }
        }
        return false
    }
    
    private func attemptStorageRecovery() -> Bool {
        do {
            // Try to compact Core Data store
            try coreDataManager.compactStore()
            errorService.clearError()
            return true
        } catch {
            return false
        }
    }
}

// MARK: - Error Analytics

extension ErrorIntegration {
    
    func trackErrorMetrics() {
        // In production, this would send error metrics to analytics
        let errorCounts = getErrorCounts()
        print("📊 Error metrics: \(errorCounts)")
    }
    
    private func getErrorCounts() -> [String: Int] {
        // This would track error frequencies
        return [
            "permission_errors": 0,
            "blocking_errors": 0,
            "data_errors": 0,
            "network_errors": 0,
            "unknown_errors": 0
        ]
    }
}

// MARK: - Notification Extensions

extension Notification.Name {
    static let usageTrackingError = Notification.Name("usageTrackingError")
    static let appBlockingError = Notification.Name("appBlockingError")
    static let permissionStatusChanged = Notification.Name("permissionStatusChanged")
    static let dataCorruptionDetected = Notification.Name("dataCorruptionDetected")
    static let networkError = Notification.Name("networkError")
}

// MARK: - Error Context

struct ErrorContext {
    let service: String
    let operation: String
    let timestamp: Date
    let additionalInfo: [String: Any]
    
    init(service: String, operation: String, additionalInfo: [String: Any] = [:]) {
        self.service = service
        self.operation = operation
        self.timestamp = Date()
        self.additionalInfo = additionalInfo
    }
}

// MARK: - Error Reporting Helper

class ErrorReporter {
    static let shared = ErrorReporter()
    private let errorIntegration = ErrorIntegration()
    
    private init() {}
    
    func report(_ error: Error, context: ErrorContext) {
        let contextString = "\(context.service).\(context.operation)"
        errorIntegration.reportError(error, context: contextString)
        
        // Log additional context
        print("🔍 Error context: \(context)")
    }
    
    func reportPermissionDenied(_ type: ErrorHandlingService.PermissionType) {
        errorIntegration.reportPermissionError(type, denied: true)
    }
    
    func reportPermissionRevoked(_ type: ErrorHandlingService.PermissionType) {
        errorIntegration.reportPermissionError(type, denied: false)
    }
}

File: android/app/src/main/java/com/awaytime/app/service/ErrorIntegration.kt
package com.awaytime.app.service

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import java.util.*

/**
 * Integration service that connects error handling with other app services
 */
class ErrorIntegration(private val context: Context) : ViewModel() {

    private val errorService = ErrorHandlingService(context)
    private val usageTrackingService = UsageTrackingService(context)
    private val appBlockingService = AppBlockingService(context)
    private val goalTrackingService = GoalTrackingService(context)
    private val permissionService = PermissionService(context)

    // Event flows for error reporting
    private val _errorReportFlow = MutableSharedFlow<ErrorReport>(replay = 0)
    val errorReportFlow: SharedFlow<ErrorReport> = _errorReportFlow.asSharedFlow()

    init {
        setupErrorIntegration()
        startErrorMonitoring()
    }

    private fun setupErrorIntegration() {
        // Initialize global error handler
        GlobalErrorHandler.initialize(context)

        // Start error monitoring
        errorService.startErrorMonitoring()
    }

    private fun startErrorMonitoring() {
        // Monitor usage tracking health
        monitorUsageTracking()

        // Monitor app blocking health
        monitorAppBlocking()

        // Monitor permission status
        monitorPermissions()

        // Monitor data integrity
        monitorDataIntegrity()
    }

    private fun monitorUsageTracking() {
        viewModelScope.launch {
            while (true) {
                delay(60000) // Check every minute
                checkUsageTrackingHealth()
            }
        }
    }

    private fun checkUsageTrackingHealth() {
        viewModelScope.launch {
            try {
                usageTrackingService.getCurrentUsage()
            } catch (e: Exception) {
                errorService.handleUsageStatsError(e)
                reportError(e, "UsageTracking", "getCurrentUsage")
            }
        }
    }

    private fun monitorAppBlocking() {
        viewModelScope.launch {
            while (true) {
                delay(120000) // Check every 2 minutes
                checkAppBlockingHealth()
            }
        }
    }

    private fun checkAppBlockingHealth() {
        viewModelScope.launch {
            try {
                appBlockingService.validateBlockingStatus()
            } catch (e: Exception) {
                errorService.handleAccessibilityError(e)
                reportError(e, "AppBlocking", "validateBlockingStatus")
            }
        }
    }

    private fun monitorPermissions() {
        viewModelScope.launch {
            while (true) {
                delay(300000) // Check every 5 minutes instead of 30 seconds
                checkPermissionStatus()
            }
        }
    }

    private fun checkPermissionStatus() {
        viewModelScope.launch {
            try {
                if (!permissionService.hasUsageStatsPermission()) {
                    errorService.handlePermissionError(
                        ErrorHandlingService.PermissionType.USAGE_STATS,
                        denied = false
                    )
                }

                if (!permissionService.hasAccessibilityPermission()) {
                    errorService.handlePermissionError(
                        ErrorHandlingService.PermissionType.ACCESSIBILITY,
                        denied = false
                    )
                }

                // Only check notification permission, skip overlay permission 
                // as it's not critical for core functionality
                if (!permissionService.hasNotificationPermission()) {
                    // Don't show error notifications for non-critical permissions
                    println("⚠️ Notification permission not granted, but app can continue")
                }
            } catch (e: Exception) {
                reportError(e, "PermissionMonitoring", "checkPermissionStatus")
            }
        }
    }

    private fun monitorDataIntegrity() {
        viewModelScope.launch {
            while (true) {
                delay(300000) // Check every 5 minutes
                checkDataIntegrity()
            }
        }
    }

    private fun checkDataIntegrity() {
        viewModelScope.launch {
            try {
                // Test basic data operations
                val prefs = context.getSharedPreferences("awaytime_prefs", Context.MODE_PRIVATE)
                prefs.getString("test_key", null) // Test read access

                goalTrackingService.getCurrentGoals() // Test goal data
                usageTrackingService.getTodayUsage() // Test usage data
            } catch (e: Exception) {
                errorService.handleStorageError(e)
                reportError(e, "DataIntegrity", "checkDataIntegrity")
            }
        }
    }

    // MARK: - Public Interface

    fun getErrorService(): ErrorHandlingService {
        return errorService
    }

    fun reportError(error: Exception, service: String, operation: String) {
        val errorReport = ErrorReport(
            error = error,
            service = service,
            operation = operation,
            timestamp = System.currentTimeMillis(),
            context = getErrorContext()
        )

        viewModelScope.launch {
            _errorReportFlow.emit(errorReport)
        }

        println("🚨 Error reported from $service.$operation: ${error.message}")
        errorService.handleUnknownError(error)
    }

    fun reportPermissionError(type: ErrorHandlingService.PermissionType, denied: Boolean = true) {
        errorService.handlePermissionError(type, denied)
    }

    fun reportBlockingFailure(appName: String) {
        errorService.handleBlockingError(appName)
    }

    fun reportDataCorruption() {
        errorService.handleDataCorruption()
    }

    fun reportNetworkError() {
        errorService.handleNetworkError()
    }

    private fun getErrorContext(): Map<String, String> {
        return mapOf(
            "app_version" to getAppVersion(),
            "android_version" to android.os.Build.VERSION.RELEASE,
            "device_model" to android.os.Build.MODEL,
            "available_memory" to getAvailableMemory(),
            "battery_level" to getBatteryLevel()
        )
    }

    private fun getAppVersion(): String {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, 0)
            packageInfo.versionName ?: "unknown"
        } catch (e: Exception) {
            "unknown"
        }
    }

    private fun getAvailableMemory(): String {
        return try {
            val activityManager =
                context.getSystemService(Context.ACTIVITY_SERVICE) as android.app.ActivityManager
            val memoryInfo = android.app.ActivityManager.MemoryInfo()
            activityManager.getMemoryInfo(memoryInfo)
            "${memoryInfo.availMem / 1024 / 1024} MB"
        } catch (e: Exception) {
            "unknown"
        }
    }

    private fun getBatteryLevel(): String {
        return try {
            val batteryManager =
                context.getSystemService(Context.BATTERY_SERVICE) as android.os.BatteryManager
            val batteryLevel =
                batteryManager.getIntProperty(android.os.BatteryManager.BATTERY_PROPERTY_CAPACITY)
            "$batteryLevel%"
        } catch (e: Exception) {
            "unknown"
        }
    }

    // MARK: - Auto Recovery

    fun attemptAutoRecovery(error: ErrorHandlingService.AwayTimeError): Boolean {
        return when (error) {
            is ErrorHandlingService.AwayTimeError.BlockingFailed -> {
                attemptBlockingRecovery()
            }

            is ErrorHandlingService.AwayTimeError.UsageStatsError,
            is ErrorHandlingService.AwayTimeError.AccessibilityError -> {
                attemptSystemServiceRecovery()
            }

            is ErrorHandlingService.AwayTimeError.StorageError -> {
                attemptStorageRecovery()
            }

            else -> false
        }
    }

    private fun attemptBlockingRecovery(): Boolean {
        return try {
            viewModelScope.launch {
                appBlockingService.initializeBlocking()
                errorService.clearError()
            }
            true
        } catch (e: Exception) {
            false
        }
    }

    private fun attemptSystemServiceRecovery(): Boolean {
        return try {
            viewModelScope.launch {
                // Try to restart monitoring
                usageTrackingService.startMonitoring()
                errorService.clearError()
            }
            true
        } catch (e: Exception) {
            false
        }
    }

    private fun attemptStorageRecovery(): Boolean {
        return try {
            // Try to clear temporary data
            val cacheDir = context.cacheDir
            cacheDir.listFiles()?.forEach { file ->
                if (file.isFile && file.name.startsWith("temp_")) {
                    file.delete()
                }
            }
            errorService.clearError()
            true
        } catch (e: Exception) {
            false
        }
    }

    // MARK: - Error Analytics

    fun trackErrorMetrics() {
        viewModelScope.launch {
            val errorCounts = getErrorCounts()
            println("📊 Error metrics: $errorCounts")

            // In production, this would send to analytics service
        }
    }

    private fun getErrorCounts(): Map<String, Int> {
        // This would track error frequencies from persistent storage
        return mapOf(
            "permission_errors" to 0,
            "blocking_errors" to 0,
            "data_errors" to 0,
            "network_errors" to 0,
            "unknown_errors" to 0
        )
    }
}

// MARK: - Error-Aware Service Wrappers

/**
 * Wrapper for UsageTrackingService with error handling
 */
class SafeUsageTrackingService(private val context: Context) {
    private val usageService = UsageTrackingService(context)
    private val errorIntegration = ErrorIntegration(context)

    suspend fun getCurrentUsage(): List<AppUsageData> {
        return try {
            usageService.getCurrentUsage()
        } catch (e: Exception) {
            errorIntegration.reportError(e, "UsageTracking", "getCurrentUsage")
            emptyList()
        }
    }

    fun getTodayUsage(): List<AppUsageData> {
        return try {
            usageService.getTodayUsage()
        } catch (e: Exception) {
            errorIntegration.reportError(e, "UsageTracking", "getTodayUsage")
            emptyList()
        }
    }

    fun startMonitoring() {
        try {
            kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.IO).launch {
                usageService.startMonitoring()
            }
        } catch (e: Exception) {
            errorIntegration.reportError(e, "UsageTracking", "startMonitoring")
        }
    }
}

/**
 * Wrapper for AppBlockingService with error handling
 */
class SafeAppBlockingService(private val context: Context) {
    private val blockingService = AppBlockingService(context)
    private val errorIntegration = ErrorIntegration(context)

    suspend fun blockApps(apps: List<String>) {
        try {
            blockingService.blockApps(apps)
        } catch (e: Exception) {
            // Report specific app blocking failures
            apps.forEach { app ->
                errorIntegration.reportBlockingFailure(app)
            }
        }
    }

    suspend fun unblockApps(apps: List<String>) {
        try {
            blockingService.unblockApps(apps)
        } catch (e: Exception) {
            errorIntegration.reportError(e, "AppBlocking", "unblockApps")
        }
    }

    suspend fun initializeBlocking() {
        try {
            blockingService.initializeBlocking()
        } catch (e: Exception) {
            errorIntegration.reportError(e, "AppBlocking", "initializeBlocking")
        }
    }
}

/**
 * Wrapper for GoalTrackingService with error handling
 */
class SafeGoalTrackingService(private val context: Context) {
    private val goalService = GoalTrackingService(context)
    private val errorIntegration = ErrorIntegration(context)

    fun saveGoal(goal: Goal) {
        try {
            goalService.saveGoal(goal)
        } catch (e: Exception) {
            errorIntegration.reportError(e, "GoalTracking", "saveGoal")
        }
    }

    fun getCurrentGoals(): List<Goal> {
        return try {
            goalService.getCurrentGoals()
        } catch (e: Exception) {
            errorIntegration.reportError(e, "GoalTracking", "getCurrentGoals")
            emptyList()
        }
    }

    fun updateGoalProgress(goalId: String, progress: Double) {
        try {
            goalService.updateGoalProgress(goalId, progress)
        } catch (e: Exception) {
            errorIntegration.reportError(e, "GoalTracking", "updateGoalProgress")
        }
    }
}

// MARK: - Data Classes

data class ErrorReport(
    val error: Exception,
    val service: String,
    val operation: String,
    val timestamp: Long,
    val context: Map<String, String>
)

data class ErrorContext(
    val service: String,
    val operation: String,
    val timestamp: Long = System.currentTimeMillis(),
    val additionalInfo: Map<String, Any> = emptyMap()
)

// MARK: - Error Reporter

object ErrorReporter {
    private var errorIntegration: ErrorIntegration? = null

    fun initialize(context: Context) {
        errorIntegration = ErrorIntegration(context)
    }

    fun report(error: Exception, context: ErrorContext) {
        errorIntegration?.reportError(error, context.service, context.operation)

        // Log additional context
        println("🔍 Error context: $context")
    }

    fun reportPermissionDenied(type: ErrorHandlingService.PermissionType) {
        errorIntegration?.reportPermissionError(type, denied = true)
    }

    fun reportPermissionRevoked(type: ErrorHandlingService.PermissionType) {
        errorIntegration?.reportPermissionError(type, denied = false)
    }

    fun reportBlockingFailure(appName: String) {
        errorIntegration?.reportBlockingFailure(appName)
    }

    fun reportDataCorruption() {
        errorIntegration?.reportDataCorruption()
    }

    fun reportNetworkError() {
        errorIntegration?.reportNetworkError()
    }
}

// MARK: - Extension Functions for AppBlockingService

fun AppBlockingService.validateBlockingStatus() {
    // This would check if the accessibility service is running
    // and if blocking is working correctly
    val isServiceRunning = isAccessibilityServiceEnabled()
    if (!isServiceRunning) {
        throw Exception("Accessibility service is not running")
    }
}

fun AppBlockingService.isAccessibilityServiceEnabled(): Boolean {
    // Implementation would check if the accessibility service is enabled
    val accessibilityService = AwayTimeAccessibilityService.getInstance()
    return accessibilityService != null
}